src/atlas/interpolation/method/structured/RegionalLinear2D.h

```diff
@@ -0,0 +1,76 @@
+/*
+ * (C) Copyright 2024 Meteorologisk Institutt
+ *
+ * This software is licensed under the terms of the Apache Licence Version 2.0
+ * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+#pragma once
+
+#include "atlas/interpolation/method/Method.h"
+
+#include <vector>
+
+#include "atlas/field.h"
+#include "atlas/functionspace.h"
+#include "atlas/grid/Grid.h"
+#include "atlas/interpolation/element/InterpElement.h"
+
+#include "eckit/config/Configuration.h"
+#include "eckit/mpi/Comm.h"
+
+namespace atlas {
+namespace interpolation {
+namespace method {
+
+
+class RegionalLinear2D : public Method {
+ public:
+  /// @brief   Regional linear interpolation
+  ///
+  /// @details 
+  ///
+  RegionalLinear2D(const Config& config) : Method(config), comm_(eckit::mpi::comm()) {}
+  ~RegionalLinear2D() override {}
+
+  void print(std::ostream&) const override;
+  const FunctionSpace& source() const override { return source_; }
+  const FunctionSpace& target() const override { return target_; }
+
+  void do_execute(const FieldSet& sourceFieldSet, FieldSet& targetFieldSet,
+                  Metadata& metadata) const override;
+  void do_execute(const Field& sourceField, Field& targetField,
+                  Metadata& metadata) const override;
+
+  void do_execute_adjoint(FieldSet& sourceFieldSet,
+                          const FieldSet& targetFieldSet,
+                          Metadata& metadata) const override;
+  void do_execute_adjoint(Field& sourceField, const Field& targetField,
+                          Metadata& metadata) const override;
+
+ private:
+  using Method::do_setup;
+  void do_setup(const FunctionSpace& source, const FunctionSpace& target) override;
+  void do_setup(const Grid& source, const Grid& target, const Cache&) override;
+
+  FunctionSpace source_{};
+  FunctionSpace target_{};
+
+  const eckit::mpi::Comm & comm_;
+  size_t sourceSize_;
+  std::vector<int> mpiTask_;
+  size_t targetSize_;
+  size_t sourceSendSize_;
+  size_t targetRecvSize_;
+  std::vector<int> sourceSendCounts_;
+  std::vector<int> sourceSendDispls_;
+  std::vector<int> targetRecvCounts_;
+  std::vector<int> targetRecvDispls_;
+  std::vector<size_t> sourceSendMapping_;
+  std::vector<atlas::interpolation::element::InterpElement> horInterp_;
```

Comment:
`std::vector<atlas::interpolation::element::InterpElement>` in current implementation of `InterpElement` is essentially the same as
`std::vector<std::vector<std::pair<size_t,double>>>`

The issue is `std::vector<std::vector< ... >>` because that infers multiple allocations on the heap within tight loop and slow memory access.
I'm thinking to avoid the `InterpElement` class altogether (and remove the file etc.)

What would be better is multiple arrays:
```c++
std::vector< std::array<size_t,4> > stencil_;
std::vector< std::array<double,4> > weights_;
std::vector< size_t > stencil_size_;
```
and then resize all these arrays to `targetSize_` before looping over target points, and fill the values at the correct place in these arrays, without using the temporary `operations` variable. This will also allow this loop to become OpenMP multi-threadable.

These 3 vectors are now each one large memory allocation because the memory of `std::array<type,4>` is known at compile time.
We need the `stencil_size_` array to avoid going into possibly over-allocated entries in case of linear interpolation or point value, which will almost never be the case.

Reply:
Great suggestion. For regional DA, we often interpolate from a grid to another one with exactly the same rectangular domain, and just a different cell size, so `stencil_size_` might be lower than 4 quite often.